text/1009-group-by.rst

```diff
@@ -0,0 +1,394 @@
+
+::
+
+   GROUP
+       [<alias> := ] <expr>
+
+   BY
+       [<alias1> := ] <expr>,  # define parameters to use for
+       [<alias2> := ] <expr>,  # grouping
+       ...
+       [<aliasN> := ] <expr>
+
+   [USING <alias>, ... ]       # specify which parameters will
+                               # be used to partition the set;
+                               # if unspecified, use all aliases declared in BY
+
+
+
+The aliases in the BY clause are optional only if the expression is an
+partial path reference such as ``.name``, and ``.name`` is treated as
+equivalent to ``name := .name``, except that ``name`` is not made available
+to later aliases in the ``BY`` clause.
+
+The ``USING`` clause, if omitted, is taken to be the list of all specified
+aliases in the ``BY`` clause. This means it is usually superfluous in this
+initial pre-grouping sets version, but it does allow aliases in the BY
+clause to be defined solely as a helper to later aliases.
+
+The optional alias in the subject will be bound in the ``BY`` clauses.
+
+The *behavior* of ``GROUP BY`` is to evaluate the subject expression and
+then, for each element of the set, evaluate each of the ``BY``
+expressions (which much be singletons). Then a grouping key is created
+based on the ``USING`` clause, and the objects are aggregated into groups
+with matching grouping keys.
+
+The output is reflected in an ad-hoc free object of the following form::
+
+  {
+      key: <another free object containing "alias := value" pointers>,
+      elements: <the set of all the objects with the matching key>,
+      grouping: <array of keys used in grouping>,
+  }
+
+
+
+Grouping sets
+-------------
+
+To extend this to support multiple grouping sets, we alter the syntax for the
+``USING`` clause.
+
+To do this, we generalize ``USING`` as such::
+
+  [USING <grouping-element>, ...]
+
+where ``grouping-element`` is one of::
+
+  <alias-or-list>
+  {<grouping-element>, ...}  # grouping sets
+  ROLLUP (<alias-or-list>, ...)
+  CUBE (<alias-or-list>, ...)
+
+and ``alias-or-list`` is one of::
+
+  <ident>
+  ()
+  (<ident>, ...)
+
+
+``ROLLUP`` and ``CUBE`` are basic syntactic sugar for grouping sets.
+``ROLLUP`` groups by all prefixes of a list of elements, so
+``ROLLUP (a, b, c)`` is equivalent to ``{(), (a), (a, b), (a, b, c)}``
+while ``CUBE`` considers all elements of the power set, so that
+``CUBE (a, b)`` is equivalent to ``{(), (a), (b), (a, b)}``.
+These elements can be lists of aliases, also.
+
+When grouping by a grouping set, we group by *each* element of the
+grouping set. The ``grouping`` element of the output shape indicates
+which grouping set a group is from, by listing all the keys used in
+the group, in the order they appeared in the ``BY`` clause.
+
+When there are multiple top-level ``grouping-element``s, and some are
+grouping sets, then the cartesian product of them is taken to determine
+the final grouping sets. That is ``a, {b, c}`` is equivalent to
+``{(a, b), (a, c)}``.
```

Comment:
I agree that a Cartesian product here seems complex.  I also wonder if we can drop the curly brackets altogether and make it so `grouping-element` is always either a tuple-like or a callable-like, e.g.:

```
USING
    (a, b),
    (c, d)
```

or

```
USING
   (a, b),
   CUBE(c, d)
```

Reply:
Yeah, I agree that this is fairly confusing. I did it this way because:
1. It matches SQL, and so ought to be easy to implement
2. It sort of naturally fit in with having a `USING` clause where I felt we wanted to allow a non parenthesized list at the top level, and it felt the most natural way to do that was to allow any `grouping-element` there.

I can think of two ways to restrict it:
1. Only allow multiple top-level `grouping-element`s if they are just names. (This might need to be a post-parsing check but that's fine.)
2. Make USING take a *single* `grouping-element`, which means grouping by multiple things would require parentheses. (That might be just fine, since the common case of `USING` will probably be grouping sets).

That said, I also don't hate it. To be honest, your cube/cube thing makes me kind of want to keep it, since it would be a pain to express that otherwise!

Reply:
> Overall, I still think that `USING` clause is confusing both because this name is used in other places in a very different way, 

After taking a fresh look at the design, I'm also starting to think that perhaps `USING` isn't the best name for the clause, but I don't see any way of _not_ having this clause without compromising the design in major ways.

As for the keyword, perhaps we should use `GROUPINGS`, e.g.:

```
GROUP
  Foo
BY
  a := ...,
  b := ...
GROUPINGS
  (a),
  (b),
  (a, b),
  ()
```

Reply:
Hm, though maybe it is too noisy to require singleton grouping sets to be parenthesized (that is `(a)` instead of `a`). I feel like we have to, though, with the "`GROUPINGS` is a list of grouping sets" syntax?

Reply:
I don't see a problem with always requiring parentheses.  If anything, it actually makes things clearer.  Otherwise I'd always be wondering if `GROUPINGS a, b` means `GROUPINGS (a, b)` or `GROUPINGS (a), (b)`

Reply:
I don't like `GROUPINGS`. Maybe we go with SQL's `GROUPING SETS` compound keyword?

Reply:
I updated with this changes

Reply:
`GROUPING SETS` means a slightly different thing in SQL.

Reply:
"grouping sets" always sounds needlessly verbose to me, and `GROUPINGS` matches the proposed `grouping` field in the shape

Reply:
> GROUPING SETS means a slightly different thing in SQL.

Yeah, I've just glanced over SQL docs.

Still `GROUP BY ... GROUPINGS` doesn't sound good to me. We've payed a lot of attention before trying to make sure that EdgeQL reads close to spoken English (actually the effort was usually spearheaded by Elvis).

E.g. `GROUP BY ... WITH GROUPINGS` or `GROUP BY ... USING GROUPINGS` reads better.

Reply:
Well, on the other hand we try to be as terse as possible, and in this case having to type two long keywords does not seem to be worth it to me.

Reply:
`GROUP Foo BY ... GROUPINGS CUBE(element, nowners)` 
vs `GROUP Foo BY ... WITH GROUPINGS CUBE (element, nowners)`
vs `GROUP Foo BY ... USING CUBE (element, nowners)`

Somehow, looking at the above I start to find `USING` more appealing again.

Reply:
FWIW I like `USING` 10x more than `GROUPINGS` and generally don't understand what's our problem with it. And it's not like `GROUPINGS CUBE ()` would be intimately familiar to all, people would still have to google it, just as they would google what `USING CUBE` means.

I'm strongly in favor of `USING`.

Reply:
Sure.

Reply:
How about reversing `BY` and `USING`, and still always requiring `BY`?
```
GROUP Hero BY .age, .gender;
GROUP Hero USING age_group := math::floor(.age/10) BY age_group;
GROUP Hero BY {(.age, .something), .age};
GROUP Hero BY CUBE(.x, .y);
GROUP Hero BY {CUBE(.x, .y), .z};
```

Reply:
> How about reversing `BY` and `USING`, and still always requiring `BY`?

It's a possibility, but we have to optimize for the most common/likely use case, though.  Essentially we need to decide which of the following is more common?

1. Complex grouping sets
2. Expressions in grouping keys

I think it's overwhelmingly (2), because it's common to group by some function of property (e.g. `str_lower(.first_name)` or `.age // 10`).  On the other hand, grouping sets are rarely needed outside of analytical reports.

The grouping set clause might also be necessary if you have an intermediate computation in the key clause, but DON'T want it to be a grouping key, e.g.:

```
GROUP ... BY temp := ... , x := temp, y := temp + 1 USING (x, y)
```

I feel this is marginal as well, and you can further avoid this by either using a computed in the input subject, or using a schema computed.  Another option is to add an alias convention, where aliases beginning with underscore don't get automatically included in the implicit grouping set, e.g.:

```
GROUP ... BY _temp := ... , x := _temp, y := _temp + 1
``` 

Though I feel the latter is fairly ugly and unprecedented in the language (we don't assign meanings to variable names anywhere else).

Reply:
> Essentially we need to decide which of the following is more common?

I agree with you on which one is more common. But we should not optimize solely which on is short. Also `USING` that is occurring more rarely might be more confusing, as people will never remember it and will have to look up it in the docs every time.

```
GROUP ... USING lower_name := str_lower(.first_name) BY lower_name 
```
Is good enough. We don't need to squeeze every character of it.

*(Also I'm arguing that a computable property in schema will often replace computation in group by, but it's not important)*

On the other hand, always grouping by `BY` clause is much more clear and less things to keep in mind while reading queries (and in this case missing `USING` while skimming will impact your query a lot less, and you can kinda mentally "zoom in" into `BY` clause if you care about grouping(s), and `USING` clause if you care about computations).

Reply:
> But we should not optimize solely which on is short.

+1.

I think I side with Paul; I find `GROUP .. USING .. BY` to be more readable and self-explaining than the alternative.

Reply:
So, we're back to the original `GROUP` layout essentially.  To be honest, I don't find the "rare clause" rationale particularly compelling, because I don't think that "using mandatory syntax as a teaching tool" should be a guiding principle for EdgeQL.  Relative brevity of common queries, on the other hand, was always one of those principles.

That said, I agree that `USING` being a pure alias-defining clause with `BY` specifying grouping sets is cleaner design overall.  At the same time, I think we should consider de-emphasizing `USING` by moving key expressions into the subject shape, e.g.:

```
GROUP User { lower_name := str_lower(.first_name) } BY .lower_name 
```

instead of

```
GROUP User USING lower_name := str_lower(.first_name) BY lower_name 
```

This is purely a stylistic choice, but I think the former is better aligned with how we use computeds in `SELECT` and would require less mental switcheroo.

Reply:
I'm not bothered at all by USING as a name, and I don't find the "rare clause" stuff persuasive.

I am sympathetic to Paul's point that it is more natural to always have it be `BY` that is being grouped by.

If we go that route, though, we need to reconsider the syntax for grouping sets, because I don't think that we should require parens in basic single grouping set situations.

I propose that we return to my original syntax for grouping sets, using `{ ... }` as delimiters, but with the restriction that if there are multiple top-level `<grouping-element>`s specified, none of them can be grouping sets. That lets us write `BY .element, .cost` while prohibiting stuff like `BY .element, CUBE(a, b, c), {foo, bar}` that people found objectionable.

Reply:
I don't like the curly braces thing.  It's the same thing as `CUBE` and `ROLLUP`, which are function-like, but uses entirely different syntax.  I think it should also be an explicit keyword, e.g. `BY CUBE(a, b, c), GROUPINGS((a, b), (c, d))`

Reply:
I'm also wondering if we should just copy SQL's thing entirely, including the `GROUPING SETS` keyword, sublists and nesting of grouping sets.

Reply:
The initial proposal is exactly SQL's thing copied entirely, with the exception of the `{` `}` syntax.

I don't agree that `GROUPING SETS` is the same thing as `CUBE` and `ROLLUP`. I think it is a more primitive notion for which it is reasonable to have a dedicated syntax, and `GROUPING SETS` feels very cumbersome to me. It's nine characters longer than `CUBE` despite being a more primitive and more common operation.

I see your point, though, and am totally uninterested in dying on this hill. I'm fine with just copying SQL's thing entirely.

I am also happy to follow SQL and allow multiple `<grouping-elements>` at the top level also. It's a little confusing but not that confusing, and allows expressing things otherwise a pain to express.

Reply:
OK, so let's pretend that `BY` is actually an expression of a special data type that we'll call `reftuple` (a tuple of singleton set references, opaque to the type system).  So, the type of `BY` would be `SET OF reftuple`.  Then, let's see how we can use normal EdgeQL expression logic to produce said set.  `cube` and `rollup` can be defined as functions returning `SET OF reftuple`, e.g. `cube(reftuple) -> set of reftuple`.  Then, `{ <reftuple>, ... }` is just a normal set literal.  Additionally, for convenience, both the `BY` clause itself and the reftuple-returning functions have variadic overloads (`cube(variadic r: ref) -> set of reftuple`) so `BY a, b, c` is the same as `BY (a, b, c)`.  Finally, reftuples are always flat, so `(a, (b, c))` turns into `(a, b, c)`, which helps with the Cartesian product when making reftuples out of reftuples.

So, I guess, from this formulation curly braces and the cross product make perfect sense :-)

Reply:
Actually, `cube(a, b, c)` and `cube((a, b, c))` aren't the same thing, so we can instead say that individual refs are singleton reftuples, and there's only a single variadic definition for the functions.

Reply:
It looks like if we want to allow a trailing comma in the USING clause (we do), we need to make `BY` a reserved keyword, or else we can't distinguish at one token of lookahead between the `BY` indicating the start of the `BY` clause or it being the name of an alias we are about to define.

That shouldn't be a problem, really, though?

Reply:
Not a problem, we should make it future-reserved in v1 now.

Reply:
@elprans, all you have said about reftuples makes sense. But it's not easy to understand.

This is a bit orthogonal to the original thread, but can we make a special cross-product operator?
```
GROUP BY .a, .b  # just pair of fields
GROUP BY .division @ ROLLUP(.department, .role)
```
This at least gives you a hint that you should look for `@` operator in documentation rather that finding out that comma meaning is different than what you have thought it is.

Reply:
> comma meaning is different than what you have thought it is

But comma always means a Cartesian product in tuples.  For example:

```
SELECT ('foo', array_unpack(['a', 'b'])) # {('foo', 'a'), ('foo', 'b')}
```

> `.a, .b  # just pair of fields`

It isn't just a pair of fields, it's a tuple, we just omit parentheses to avoid clutter.

Reply:
While I agree with you Elvis, Paul has a point.

In EdgeQL tuples are created with *both* parentheses and commas. In Python it's just commas, so `a = 1,` means that `a` will ref a one-element `(1,)` tuple. In EdgeQL, `select 1,` is a syntax error. So yeah, we can say `.a, .b` would form a tuple, but it might not be immediately apparent to users that that's the case. And maybe that's OK.

Reply:
A `tuple` here isn't just a literal tuple.  Function arguments are also tuples and clauses are essentially functions, some with custom syntax for practical reasons (e.g `ORDER BY`).

Reply:
> But comma always means a Cartesian product in tuples.

Yes I remember that. And I also remember how confusing it is. It's also not comma producing cartesian product, but any binary operator on scalars. So it's not exactly the same thing.

And also cartesian product "produced by tuple" is orthogonal to one in fields. Because with normal tuples cartesian product is produced on values rather than on set of fields. 

> > .a, .b # just pair of fields
>
> It isn't just a pair of fields, it's a tuple, we just omit parentheses to avoid clutter.

Well, it's not a tuple, because (a) it contains references of fields and tuple contains values and also (b) produces a free shape in result (these are the reasons why it's called "list" in RFC, right?). But this terminology discussion is irrelevant.

> A tuple here isn't just a literal tuple. Function arguments are also tuples and clauses are essentially functions, some with custom syntax for practical reasons (e.g ORDER BY).

This is even more confusing. If `ORDER BY` clause is a tuple, you should be able to do `WITH x := (Hero.gender, Hero.age DESC) SELECT Hero ORDER BY x` But as far as I understand you can't do that. Even if we ever do that syntax we don't have to explain that upfront for the new users. It's much easier to think of it as a separate clause with a dedicated syntax.